In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('Kangaroo.csv') # Load a CSV

In [3]:
df.head()

Unnamed: 0          id                                                url  \
0           0  20661494.0  https://www.immoweb.be/en/classified/apartment...   
1           1  20647642.0  https://www.immoweb.be/en/classified/house/for...   
2           2  20644816.0  https://www.immoweb.be/en/classified/apartment...   
3           3  20659813.0  https://www.immoweb.be/en/classified/apartment...   
4           4  20633249.0  https://www.immoweb.be/en/classified/flat-stud...   

        type      subtype  bedroomCount  bathroomCount  province   locality  \
0  APARTMENT    APARTMENT           2.0            1.0  Brussels  Etterbeek   
1      HOUSE        HOUSE           4.0            2.0  Brussels  Etterbeek   
2  APARTMENT    APARTMENT           2.0            1.0  Brussels  Etterbeek   
3  APARTMENT    APARTMENT           2.0            2.0  Brussels  Bruxelles   
4  APARTMENT  FLAT_STUDIO           1.0            1.0  Brussels  Etterbeek   

   postCode  ...  hasOffice  toiletCount  hasSwimmingPool hasFireplace  \
0      1040  ...        NaN          1.0              NaN          NaN   
1      1040  ...       True          3.0              NaN          NaN   
2      1040  ...        NaN          1.0              NaN          NaN   
3      1040  ...        NaN          2.0              NaN          NaN   
4      1040  ...        NaN          1.0              NaN          NaN   

  hasTerrace terraceSurface  terraceOrientation accessibleDisabledPeople  \
0       True            7.0                 NaN                      NaN   
1       True            1.0                 NaN                      NaN   
2       True            2.0                 NaN                      NaN   
3       True            3.0                 NaN                      NaN   
4        NaN            NaN                 NaN                      NaN   

  epcScore     price  
0        C  399000.0  
1        E  895000.0  
2        F  465000.0  
3        B  590000.0  
4        E  289000.0  

[5 rows x 53 columns]

In [4]:
df.shape

(80368, 53)

In [5]:
df.columns


Index(['Unnamed: 0', 'id', 'url', 'type', 'subtype', 'bedroomCount',
       'bathroomCount', 'province', 'locality', 'postCode', 'habitableSurface',
       'roomCount', 'monthlyCost', 'hasAttic', 'hasBasement',
       'hasDressingRoom', 'diningRoomSurface', 'hasDiningRoom',
       'buildingCondition', 'buildingConstructionYear', 'facedeCount',
       'floorCount', 'streetFacadeWidth', 'hasLift', 'floodZoneType',
       'heatingType', 'hasHeatPump', 'hasPhotovoltaicPanels',
       'hasThermicPanels', 'kitchenSurface', 'kitchenType', 'landSurface',
       'hasLivingRoom', 'livingRoomSurface', 'hasBalcony', 'hasGarden',
       'gardenSurface', 'gardenOrientation', 'parkingCountIndoor',
       'parkingCountOutdoor', 'hasAirConditioning', 'hasArmoredDoor',
       'hasVisiophone', 'hasOffice', 'toiletCount', 'hasSwimmingPool',
       'hasFireplace', 'hasTerrace', 'terraceSurface', 'terraceOrientation',
       'accessibleDisabledPeople', 'epcScore', 'price'],
      dtype='object')

In [6]:
df_columns = df.columns.tolist()
df_columns

['Unnamed: 0',
 'id',
 'url',
 'type',
 'subtype',
 'bedroomCount',
 'bathroomCount',
 'province',
 'locality',
 'postCode',
 'habitableSurface',
 'roomCount',
 'monthlyCost',
 'hasAttic',
 'hasBasement',
 'hasDressingRoom',
 'diningRoomSurface',
 'hasDiningRoom',
 'buildingCondition',
 'buildingConstructionYear',
 'facedeCount',
 'floorCount',
 'streetFacadeWidth',
 'hasLift',
 'floodZoneType',
 'heatingType',
 'hasHeatPump',
 'hasPhotovoltaicPanels',
 'hasThermicPanels',
 'kitchenSurface',
 'kitchenType',
 'landSurface',
 'hasLivingRoom',
 'livingRoomSurface',
 'hasBalcony',
 'hasGarden',
 'gardenSurface',
 'gardenOrientation',
 'parkingCountIndoor',
 'parkingCountOutdoor',
 'hasAirConditioning',
 'hasArmoredDoor',
 'hasVisiophone',
 'hasOffice',
 'toiletCount',
 'hasSwimmingPool',
 'hasFireplace',
 'hasTerrace',
 'terraceSurface',
 'terraceOrientation',
 'accessibleDisabledPeople',
 'epcScore',
 'price']

In [7]:
df.info() # Summary info of the data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80368 entries, 0 to 80367
Data columns (total 53 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                80368 non-null  int64  
 1   id                        80368 non-null  float64
 2   url                       80368 non-null  object 
 3   type                      80368 non-null  object 
 4   subtype                   80368 non-null  object 
 5   bedroomCount              73558 non-null  float64
 6   bathroomCount             66672 non-null  float64
 7   province                  80368 non-null  object 
 8   locality                  80368 non-null  object 
 9   postCode                  80368 non-null  int64  
 10  habitableSurface          67783 non-null  float64
 11  roomCount                 21948 non-null  float64
 12  monthlyCost               0 non-null      float64
 13  hasAttic                  12516 non-null  object 
 14  hasBas

In [8]:
# Check for missing values
df.isnull().sum()

Unnamed: 0                      0
id                              0
url                             0
type                            0
subtype                         0
bedroomCount                 6810
bathroomCount               13696
province                        0
locality                        0
postCode                        0
habitableSurface            12585
roomCount                   58420
monthlyCost                 80368
hasAttic                    67852
hasBasement                 51053
hasDressingRoom             77740
diningRoomSurface           73467
hasDiningRoom               66227
buildingCondition           22376
buildingConstructionYear    31262
facedeCount                 27129
floorCount                  42822
streetFacadeWidth           64858
hasLift                     61324
floodZoneType               35902
heatingType                 33215
hasHeatPump                 72895
hasPhotovoltaicPanels       72320
hasThermicPanels            77256
kitchenSurface

### sumary
- Shape: 80,368 x 53
- Some columns are likely unnecessary: url, id?, unnamed, 
- Many columns misses a lot information:
100%: monthlyCost, hasBalcony, accessibleDisabledPeople

## Data cleaning
1. Remove duplicate : No duplicates in data set
2. Remove irrelevant columns: URL, Unnamed:0, 
3. Remove columns missing a lot info


In [ ]:
# df.duplicated().sum() # Check for duplicate rows
df.drop_duplicates(subset=['id'], keep='last', inplace=True)


np.int64(0)

In [ ]:
# 1. Drop rows with NaN in price column
df.dropna(subset=['price'], how='all', inplace=True)
df.shape

(80368, 51)

In [10]:
# 2. Drop irrelevant columns
df_cleaned = df.drop(columns=["Unnamed: 0", "url"])
df_cleaned.shape
df_cleaned.head()


id       type      subtype  bedroomCount  bathroomCount  province  \
0  20661494.0  APARTMENT    APARTMENT           2.0            1.0  Brussels   
1  20647642.0      HOUSE        HOUSE           4.0            2.0  Brussels   
2  20644816.0  APARTMENT    APARTMENT           2.0            1.0  Brussels   
3  20659813.0  APARTMENT    APARTMENT           2.0            2.0  Brussels   
4  20633249.0  APARTMENT  FLAT_STUDIO           1.0            1.0  Brussels   

    locality  postCode  habitableSurface  roomCount  ...  hasOffice  \
0  Etterbeek      1040             100.0        NaN  ...        NaN   
1  Etterbeek      1040             270.0       12.0  ...       True   
2  Etterbeek      1040              87.0        NaN  ...        NaN   
3  Bruxelles      1040             104.0        NaN  ...        NaN   
4  Etterbeek      1040              71.0        NaN  ...        NaN   

  toiletCount hasSwimmingPool hasFireplace  hasTerrace terraceSurface  \
0         1.0             NaN          NaN        True            7.0   
1         3.0             NaN          NaN        True            1.0   
2         1.0             NaN          NaN        True            2.0   
3         2.0             NaN          NaN        True            3.0   
4         1.0             NaN          NaN         NaN            NaN   

  terraceOrientation  accessibleDisabledPeople  epcScore     price  
0                NaN                       NaN         C  399000.0  
1                NaN                       NaN         E  895000.0  
2                NaN                       NaN         F  465000.0  
3                NaN                       NaN         B  590000.0  
4                NaN                       NaN         E  289000.0  

[5 rows x 51 columns]

In [11]:
#3. Trim Whitespace in Strings

str_cols = df_cleaned.select_dtypes(include='object').columns #Selects columns that are of type string or mixed object
for col in str_cols:
    df_cleaned[col] = df_cleaned[col].apply(lambda x: x.strip() if isinstance(x, str) else x)

df_cleaned.describe(include='all') # Summary statistics of the data



id   type subtype  bedroomCount  bathroomCount  \
count   8.036800e+04  80368   80368  73558.000000   66672.000000   
unique           NaN      4      26           NaN            NaN   
top              NaN  HOUSE   HOUSE           NaN            NaN   
freq             NaN  45234   33977           NaN            NaN   
mean    1.952922e+07    NaN     NaN      3.008877       1.427121   
std     2.897688e+06    NaN     NaN      1.833185       4.556189   
min     1.882546e+06    NaN     NaN      1.000000       1.000000   
25%     2.035055e+07    NaN     NaN      2.000000       1.000000   
50%     2.052985e+07    NaN     NaN      3.000000       1.000000   
75%     2.061494e+07    NaN     NaN      4.000000       2.000000   
max     2.066439e+07    NaN     NaN    100.000000    1146.000000   

             province   locality      postCode  habitableSurface  \
count           80368      80368  80368.000000      67783.000000   
unique             11       5470           NaN               NaN   
top     West Flanders  Antwerpen           NaN               NaN   
freq            15515       1665           NaN               NaN   
mean              NaN        NaN   5266.587199        178.499993   
std               NaN        NaN   3045.698904        820.878936   
min               NaN        NaN   1000.000000          6.000000   
25%               NaN        NaN   2330.000000         93.000000   
50%               NaN        NaN   4830.000000        140.000000   
75%               NaN        NaN   8420.000000        200.000000   
max               NaN        NaN   9992.000000     185347.000000   

           roomCount  ...  hasOffice  toiletCount hasSwimmingPool  \
count   21948.000000  ...      10387  55094.00000            1816   
unique           NaN  ...          1          NaN               1   
top              NaN  ...       True          NaN            True   
freq             NaN  ...      10387          NaN            1816   
mean        7.110443  ...        NaN      1.80624             NaN   
std         5.824088  ...        NaN      8.41452             NaN   
min         1.000000  ...        NaN      1.00000             NaN   
25%         1.000000  ...        NaN      1.00000             NaN   
50%         6.000000  ...        NaN      2.00000             NaN   
75%        11.000000  ...        NaN      2.00000             NaN   
max        68.000000  ...        NaN   1958.00000             NaN   

       hasFireplace  hasTerrace terraceSurface terraceOrientation  \
count          3044       47429   28599.000000              11262   
unique            1           1            NaN                  8   
top            True        True            NaN              SOUTH   
freq           3044       47429            NaN               3143   
mean            NaN         NaN      24.512710                NaN   
std             NaN         NaN      43.244643                NaN   
min             NaN         NaN       1.000000                NaN   
25%             NaN         NaN       8.000000                NaN   
50%             NaN         NaN      15.000000                NaN   
75%             NaN         NaN      29.000000                NaN   
max             NaN         NaN    2921.000000                NaN   

        accessibleDisabledPeople  epcScore         price  
count                        0.0     65391  7.637000e+04  
unique                       NaN        20           NaN  
top                          NaN         B           NaN  
freq                         NaN     13243           NaN  
mean                         NaN       NaN  4.476061e+05  
std                          NaN       NaN  5.115757e+05  
min                          NaN       NaN  3.141000e+03  
25%                          NaN       NaN  2.300000e+05  
50%                          NaN       NaN  3.299000e+05  
75%                          NaN       NaN  4.790000e+05  
max                          NaN       NaN  1.500000e+07  

[11 rows x 51 col

In [ ]:
# check for missing values 
#  Percentage of missing values
missing_percent = df_cleaned.isnull().mean().sort_values(ascending=False) * 100
missing_percent

id                          0.000000
type                        0.000000
subtype                     0.000000
bedroomCount                0.084735
bathroomCount               0.170416
province                    0.000000
locality                    0.000000
postCode                    0.000000
habitableSurface            0.156592
roomCount                   0.726906
monthlyCost                 1.000000
hasAttic                    0.844266
hasBasement                 0.635240
hasDressingRoom             0.967300
diningRoomSurface           0.914132
hasDiningRoom               0.824047
buildingCondition           0.278419
buildingConstructionYear    0.388986
facedeCount                 0.337560
floorCount                  0.532824
streetFacadeWidth           0.807013
hasLift                     0.763040
floodZoneType               0.446720
heatingType                 0.413286
hasHeatPump                 0.907015
hasPhotovoltaicPanels       0.899861
hasThermicPanels            0.961278
k

In [ ]:
# 4. Handle Missing Values
# Drop columns with over 70% missing
df_cleaned = df_cleaned.drop(columns=missing_percent[missing_percent > 70].index)



In [ ]:
df_cleaned.head()

In [ ]:
df_cleaned.isnull().mean().sort_values(ascending=False) * 100

In [ ]:
df_cleaned.columns.tolist()

### 2. Data Analysis
- Linear correlation: 

In [ ]:
df_cleaned.shape

In [ ]:
# Compute Pearson correlation of every numeric column with price
corr_with_price = df_cleaned.corr(numeric_only=True)['price'].sort_values(ascending=False)

# View top positive and negative correlations
print("Top positive correlations:\n", corr_with_price.head(10))
print("\nTop negative correlations:\n", corr_with_price.tail(10))

In [ ]:
numeric_df = df_cleaned.select_dtypes(include=[np.number])

# 2) Compute the correlation matrix
corr_matrix = numeric_df.corr()

In [ ]:
fig, ax = plt.subplots(figsize=(20,16))
cax = ax.matshow(corr_matrix.values, cmap='coolwarm', vmin=-1, vmax=1)
fig.colorbar(cax, shrink=0.8)

# tick labels
ticks = np.arange(len(corr_matrix.columns))
ax.set_xticks(ticks)
ax.set_yticks(ticks)
ax.set_xticklabels(corr_matrix.columns, rotation=90)
ax.set_yticklabels(corr_matrix.columns)

# manual annotation loop
for (i, j), val in np.ndenumerate(corr_matrix.values):
    ax.text(j, i, f"{val:.2f}", ha='center', va='center', fontsize=7)

plt.title("Fully Annotated Correlation Matrix", pad=20)
plt.tight_layout()
plt.show()

In [ ]:
most_influential = corr_with_price.abs().sort_values(ascending=False).index[1:6]
least_influential = corr_with_price.abs().sort_values(ascending=True).index[:5]

print("Top 5 most influential:", most_influential.tolist())
print("Top 5 least influential:", least_influential.tolist())

In [ ]:
# Quantitative = numeric types
quant_cols = df_cleaned.select_dtypes(include=['number']).columns.tolist()
# Qualitative = object / category types
qual_cols = df_cleaned.select_dtypes(include=['object', 'category']).columns.tolist()

print(f"Quantitative ({len(quant_cols)}): {quant_cols}")
print(f"Qualitative ({len(qual_cols)}): {qual_cols}")

In [ ]:
missing_count = df_cleaned.isnull().sum()
missing_pct   = (missing_count / len(df_cleaned) * 100).round(2)

missing_summary = pd.DataFrame({
    'missing_count': missing_count,
    'missing_pct'  : missing_pct
}).sort_values('missing_pct', ascending=False)

print(missing_summary.head(10))  # top 10 most-missing

In [ ]:
#Histogram of price
plt.hist(df_cleaned['price'], bins=50)
plt.title("Price Distribution")
plt.xlabel("Price")
plt.ylabel("Count")
plt.show()

In [ ]:
# Scatter price vs. habitableSurface
plt.scatter(df_cleaned['habitableSurface'], df_cleaned['price'], alpha=0.3)
plt.title("Price vs. Habitable Surface")
plt.xlabel("Surface (m²)")
plt.ylabel("Price")
plt.show()

In [ ]:
#Boxplot to spot outliers in price by province
sns.boxplot(x='province', y='price', data=df_cleaned)
plt.xticks(rotation=45)
plt.title("Price by Province")
plt.show()

### Non-linear correlation
- Spearman Rank Correlation

In [ ]:
df_cleaned.corr(method='spearman')['price'].sort_values(ascending=False)

In [ ]:
spearman_corr = numeric_df.corr(method='spearman')
spearman_price_corr = spearman_corr['price'].drop('price').sort_values(ascending=False)

# Visualize
plt.figure(figsize=(10, 6))
sns.barplot(x=spearman_price_corr.values, y=spearman_price_corr.index, palette='viridis')
plt.title("Spearman Correlation with Price")
plt.xlabel("Spearman correlation coefficient")
plt.tight_layout()
plt.show()

In [ ]:
#Scatterplots for Visual Inspection
# For a few features (pick ones with low Pearson but possible importance)
features_to_check = ['buildingConstructionYear', 'terraceSurface', 'landSurface']

for feature in features_to_check:
    plt.figure(figsize=(6, 4))
    sns.scatterplot(x=numeric_df[feature], y=numeric_df['price'])
    plt.title(f"Scatterplot: {feature} vs Price")
    plt.xlabel(feature)
    plt.ylabel('Price')
    plt.tight_layout()
    plt.show()

In [ ]:
df_cleaned['bedroomCount'].value_counts()
#df.iloc[[df['bedroomCount'].idxmax()]]

In [ ]:
df_cleaned['bathroomCount'].value_counts()



In [ ]:
df_cleaned.iloc[[df_cleaned['bathroomCount'].idxmax()]]